### Table of Contents:
Imports: Imported libraries  
Dataframes: Dataframes to be used  
1: Delete missing values and duplicates from products dataframe   
2: Statistics for products dataframe   
3: Check for mixed-type values in orders dataframe   
4: Report for mixed-type data check  
5: Missing values for orders dataframe   
6: Adressing missing values from orders dataframe   
7: Check for duplicates from orders dataframe   
8: Report for duplicates check  
9: Export products dataframe and orders dataframe

# Imports

In [1]:
import pandas
import numpy
import os

# Dataframes

In [2]:
df_orders = pandas.read_csv(r"C:\Users\justi\Downloads\Instacart Basket Analysis\2. Data\Prepared\orders_wrangled.csv")
df_products = pandas.read_csv(r"C:\Users\justi\Downloads\Instacart Basket Analysis\2. Data\Original\products.csv")

# 1

In [3]:
# Show number of missing values in each column
df_products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [4]:
# Show all records with a missing product name
df_products[df_products['product_name'].isnull() == True]

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [5]:
# Show number of rows before consistency checks
df_products.shape

(49693, 5)

In [6]:
# Create new products dataframe without records containing missing values
df_products_consistent = df_products[df_products['product_name'].isnull() == False]
df_products_consistent.shape

(49677, 5)

In [7]:
# Update dataframe with duplicates removed
df_products_consistent = df_products_consistent.drop_duplicates()
df_products_consistent.shape

(49672, 5)

# 2

In [8]:
df_products_consistent.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The prices column should be further investigated, because the max is too high.

# 3

In [9]:
# Check for mixed-type values
for col in df_orders.columns.tolist():
  mixtype = (df_orders[[col]].applymap(type) != df_orders[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_orders[mixtype]) > 0:
    print (col)

# 4

No mixed-type data found

# 5

In [10]:
# Check for missing values in the orders dataframe
df_orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in the days_since_prior_order_column. These are probably first-time orders.

# 6

Nothing was done with missing values under the assumption that values are missing because they are first-time orders.

# 7

In [38]:
# Check for duplicates
df_orders[df_orders.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, over_30 days_since_last_order]
Index: []

If there were duplicates found, a new dataframe without the duplicates could be created.

# 8

There were no duplicated found.

# 9

In [39]:
# Export dataframes
df_products_consistent.to_csv(r"C:\Users\justi\Downloads\Instacart Basket Analysis\2. Data\Prepared\products_prepared.csv")
df_orders.to_csv(r"C:\Users\justi\Downloads\Instacart Basket Analysis\2. Data\Prepared\orders_prepared.csv")